arrows/core/video_input_arf.cxx

```diff
@@ -0,0 +1,558 @@
+bool
+video_input_arf
+::seekable() const
+{
+  return true;
+}
+
+// ------------------------------------------------------------------
+size_t
+video_input_arf
+::num_frames() const
+{
+  return d->num_frames;
+}
+
+// ------------------------------------------------------------------
+bool
+video_input_arf
+::next_frame(kv::timestamp& ts,   // returns timestamp
+              VITAL_UNUSED uint32_t     timeout ) // not supported
+{
+  // reset current timestamp
+  ts = kv::timestamp();
+
+  // Check for at end of video
+  if ( this->end_of_video() )
+  {
+    return false;
+  }
+
+  ++d->frame_number;
+
+  // Check for at end of video
+  if ( this->end_of_video() )
+  {
+    return false;
+  }
+
+  // Return timestamp
+  ts = this->frame_timestamp();
+
+  return true;
+}
+
+// ------------------------------------------------------------------
+bool
+video_input_arf
+::seek_frame(kv::timestamp& ts,   // returns timestamp
+             kv::timestamp::frame_t frame_number,
+             VITAL_UNUSED uint32_t  timeout )
+{
+  // reset current timestamp
+  ts = kv::timestamp();
+
+  // Check if requested frame exists
+  if (frame_number > d->num_frames || frame_number <= 0)
+  {
+    return false;
+  }
+
+  d->frame_number = frame_number;
+
+  // Return timestamp
+  ts = this->frame_timestamp();
+
+  return true;
+}
+
+// ------------------------------------------------------------------
+kv::timestamp
+video_input_arf
+::frame_timestamp() const
+{
+  // Check for at end of video
+  if ( this->end_of_video() )
+  {
+    return {};
+  }
+
+  kv::timestamp ts;
+
+  ts.set_frame( d->frame_number );
+
+  return ts;
+}
+
+// ------------------------------------------------------------------
+kv::image_container_sptr
+video_input_arf
+::frame_image()
+{
+  if (!good())
+    return nullptr;
+
+  auto img_bytes = std::make_shared<kv::image_memory>(d->img_size);
+
+  // TODO Questions if using the GAP_SIZE is correct or not..
+  fseek(d->file, d->offset + (d->frame_number * (d->img_size + GAP_SIZE)), SEEK_SET);
+  fread(img_bytes->data(), 1, d->img_size, d->file);
+  if (d->convert_endian)
+    d->byteswap_img_bytes((unsigned char*)img_bytes->data());
+
+  kv::image img(img_bytes,img_bytes->data(),
+    d->cols, d->rows, 1, 1, d->cols, 1, d->px);
+
+  auto m = std::make_shared<kv::metadata>();
+  m->set_timestamp(this->frame_timestamp());
+  auto imc = std::make_shared<kv::simple_image_container>(img);
+  imc->set_metadata(m);
+
+  return imc;
+}
+
+// ------------------------------------------------------------------
+kv::metadata_vector
+video_input_arf
+::frame_metadata()
+{
+  kv::metadata_vector vect;
+  if ( d->metadata )
+  {
+    vect.push_back( d->metadata );
+  }
+
+  return vect;
+}
+
+// ----------------------------------------------------------------------------
+kv::metadata_map_sptr
+video_input_arf
+::metadata_map()
+{
+  return std::make_shared< kv::simple_metadata_map >(d->metadata_map);
```

Comment:
Likewise, I think you can return `nullptr` here if you don't actually have metadata.